src/sse/sse_radix4.rs

```diff
@@ -0,0 +1,507 @@
+            Sse32Butterfly::Len2(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse32Butterfly::Len4(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse32Butterfly::Len8(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse32Butterfly::Len16(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+            Sse32Butterfly::Len32(bf) => bf.perform_fft_butterfly_multi(spectrum).unwrap(),
+        }
+
+        //let end = time::Instant::now();
+        //println!("base fft: {} ns", end.duration_since(start).as_nanos());
+
+        //let start = time::Instant::now();
+        //self.base_fft.perform_fft_butterfly(spectrum); //, &mut []);
+
+        // cross-FFTs
+        let mut current_size = self.base_len * 4;
+        let mut layer_twiddles: &[__m128] = &self.twiddles;
+
+        while current_size <= signal.len() {
+            let num_rows = signal.len() / current_size;
+
+            for i in 0..num_rows {
+                butterfly_4_32(
+                    &mut spectrum[i * current_size..],
+                    layer_twiddles,
+                    current_size / 4,
+                    &self.bf4,
+                )
+            }
+
+            //skip past all the twiddle factors used in this layer
+            let twiddle_offset = (current_size * 2) / 4;
+            layer_twiddles = &layer_twiddles[twiddle_offset..];
+
+            current_size *= 4;
+        }
+        //let end = time::Instant::now();
+        //println!("cross fft: {} ns", end.duration_since(start).as_nanos());
+    }
+}
+boilerplate_fft_sse_oop!(Sse32Radix4, |this: &Sse32Radix4<_>| this.len);
+
+// after testing an iterative bit reversal algorithm, this recursive algorithm
+// was almost an order of magnitude faster at setting up
+fn prepare_radix4_32<T: FftNum>(
+    size: usize,
+    base_len: usize,
+    signal: &[Complex<T>],
+    spectrum: &mut [Complex<T>],
+    stride: usize,
+) {
+    if size == base_len {
+        unsafe {
+            for i in 0..size {
+                *spectrum.get_unchecked_mut(i) = *signal.get_unchecked(i * stride);
+            }
+        }
+    } else {
+        for i in 0..4 {
+            prepare_radix4_32(
+                size / 4,
+                base_len,
+                &signal[i * stride..],
+                &mut spectrum[i * (size / 4)..],
+                stride * 4,
+            );
+        }
+    }
+}
+
+#[target_feature(enable = "sse3")]
+unsafe fn butterfly_4_32<T: FftNum>(
+    data: &mut [Complex<T>],
+    twiddles: &[__m128],
+    num_ffts: usize,
+    bf4: &Sse32Butterfly4<T>,
+) {
+    let mut idx = 0usize;
+    let mut tw_idx = 0usize;
+    let output_slice = data.as_mut_ptr() as *mut Complex<f32>;
+    for _ in 0..num_ffts {
+        // There is no intrinsic to load the lower or upper two singles. Instead we load them as a double and the cast the __m128d to a __m128.
+        let scratch0 = _mm_load_sd(output_slice.add(idx) as *const f64);
```

Comment:
A possible next step could be loading 2 complex numbers per scratch element. Then, you'd effectively be doing 2 butterfly4's in parallel. This is where the concept of "column butterflies" came from in the AVX code. And you'd multiply the idx by 2, and loop form 0..num_ffts/2.

Ordinarily you'd have to worry about how to handle if num_ffts isn't cleanly divisible by 2, but thankfully since this is radix4, it should hopefully always be cleanly divisible (I haven't thought this all the way through, so maybe not).

Some possible benefits include:

- fewer differences between the f32 and f64 code, since it'll be largely identical besides the number of complex numbers you're loading per iteration, and the specific functions you're calling.
- you can go back to only storing 3 twiddle factors instead of 4 - because element 0 won't need to be packed with element 1
- Less data reshuffling/bookkeeping. Right now, this is executing 1 instruction per element to load data, averaged over the whole loop, whereas the proposed version would only execute 0.5 load instructions per complex element.

No pressure, you seem to be figuring things out pretty well on your own so far

Reply:
Very good suggestion! It was easy to implement and gave a nice speedup. Getting this to work took about the same amount of time to do as figuring out all that casting etc to read only two f32s into half a __m128.
This is a quick comparison of the current performance for powers of two:
![Screenshot from 2021-02-25 22-15-40](https://user-images.githubusercontent.com/6504678/109219357-cf906780-77b7-11eb-954b-d40742ee0cb0.png)